File: iosApp/iosApp/Features/Members/MemberCellCompact.swift
import SwiftUI
import Shared

/// Compact directory cell, iPhone Contacts style:
/// avatar + "FirstName **LastName**", no chevron, no secondary line.
struct MemberCellCompact: View {
    let member: RegSociModel

    var body: some View {
        HStack(spacing: 12) {
            MemberAvatar(member: member, size: 40)

            styledName
                .lineLimit(1)
                .foregroundStyle(.primary)

            Spacer(minLength: 0)
        }
        .padding(.vertical, 2)
        .contentShape(Rectangle())
    }

    /// First name regular, last name bold — matches Apple's Contacts default.
    /// Words are normalized to Title Case (the backend stores UPPERCASE).
    private var styledName: Text {
        let parts = member.name
            .split(separator: " ")
            .map { word -> String in
                guard let first = word.first else { return String(word) }
                return String(first).uppercased() + word.dropFirst().lowercased()
            }
        guard let last = parts.last else { return Text("") }
        let first = parts.dropLast().joined(separator: " ")
        if first.isEmpty {
            return Text(last).font(.body.weight(.semibold))
        }
        return Text(first).font(.body)
            + Text(" ")
            + Text(last).font(.body.weight(.semibold))
    }
}

/// Circular avatar with cached remote image and a gradient initial fallback.
struct MemberAvatar: View {
    let member: RegSociModel
    let size: CGFloat

    private var resolvedURL: URL? {
        let raw = member.image
        guard !raw.isEmpty, !isPlaceholderURL(raw) else { return nil }
        if raw.hasPrefix("http://") || raw.hasPrefix("https://") {
            return URL(string: raw)
        }
        // PocketBase serve solo i thumb DICHIARATI nella field config; richiedere
        // una size non in lista risponde 400 e il render finisce sulle iniziali.
        // `200x200` è l'unico thumb che combaci sia con la lista (40pt) sia con
        // l'hero del dettaglio (120pt). Slight under-res a 3× retina sul detail,
        // ma è la scelta sicura. Per qualità maggiore: configurare un thumb più
        // grande lato PB e aggiornarlo qui.
        return Files.url(
            collection: "members_registry",
            recordId: member.id,
            filename: raw,
            thumb: "0x100"
        )
    }

    var body: some View {
        Group {
            if let url = resolvedURL {
                CachedAsyncImage(url: url) { img in
                    img.resizable().scaledToFill()
                } placeholder: {
                    initialBubble
                }
            } else {
                initialBubble
            }
        }
        .frame(width: size, height: size)
        .clipShape(Circle())
        .overlay(
            Circle().stroke(.white.opacity(0.6), lineWidth: 1)
        )
    }

    private var initials: String {
        let parts = member.name.split(separator: " ").prefix(2)
        let chars = parts.compactMap { $0.first }
        let result = String(chars).uppercased()
        return result.isEmpty ? "?" : result
    }

    private var initialBubble: some View {
        ZStack {
            AppTheme.brandGradient
            Text(initials)
                .font(.system(size: size * 0.4, weight: .bold, design: .rounded))
                .foregroundStyle(.white)
        }
    }

    /// The legacy backend often returns this Cloud32 silhouette for users
    /// without a photo. Treat it as missing so we render the initials.
    private func isPlaceholderURL(_ url: String) -> Bool {
        url.contains("cloud32.it/Associazioni/img/Uomo-1.png")
    }
}

File: androidApp/src/main/kotlin/it/mensa/app/features/members/_components/MemberCellCompact.kt
package it.mensa.app.features.members._components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import it.mensa.app.support.FilesUrl
import it.mensa.app.ui.components.CachedAsyncImage
import it.mensa.app.ui.theme.brandGradient
import it.mensa.shared.model.RegSociModel

/**
 * Compact member cell — iPhone Contacts style.
 * Avatar (circular) + styled name (first regular, last bold).
 */
@Composable
fun MemberCellCompact(
    member: RegSociModel,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .padding(vertical = 2.dp),
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        MemberAvatar(member = member, size = 40.dp)
        StyledName(
            name = member.name,
            modifier = Modifier.weight(1f),
            style = MaterialTheme.typography.bodyMedium,
        )
    }
}

/** Circular avatar with cached remote image and gradient-initial fallback. */
@Composable
fun MemberAvatar(
    member: RegSociModel,
    size: androidx.compose.ui.unit.Dp,
    modifier: Modifier = Modifier,
) {
    // Same `0x100` thumb usato sia in lista che dall'engine Spotlight — è
    // quello sicuramente configurato lato PocketBase. Per il dettaglio è
    // disponibile `MemberHeroAvatar` che fa load progressivo 0x100 → 0x500.
    val imageUrl = remember(member) {
        val raw = member.image
        when {
            raw.isEmpty() -> null
            isPlaceholderUrl(raw) -> null
            raw.startsWith("http://") || raw.startsWith("https://") -> raw
            else -> FilesUrl.build("members_registry", member.id, raw, "0x100")
        }
    }

    val initials = remember(member.name) {
        member.name
            .split(" ")
            .take(2)
            .mapNotNull { it.firstOrNull()?.uppercaseChar() }
            .joinToString("")
            .ifEmpty { "?" }
    }

    Box(
        modifier = modifier
            .size(size)
            .clip(CircleShape),
        contentAlignment = Alignment.Center,
    ) {
        if (imageUrl != null) {
            CachedAsyncImage(
                model = imageUrl,
                contentDescription = member.name,
                modifier = Modifier.fillMaxSize(),
                contentScale = ContentScale.Crop,
            )
        } else {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .background(brandGradient),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    text = initials,
                    style = MaterialTheme.typography.bodySmall,
                    color = Color.White,
                    fontWeight = FontWeight.Bold,
                )
            }
        }
    }
}

/**
 * First name(s) regular, last word bold — Apple Contacts default.
 * Words are Title-cased (backend stores UPPERCASE).
 */
@Composable
fun StyledName(
    name: String,
    modifier: Modifier = Modifier,
    style: androidx.compose.ui.text.TextStyle = MaterialTheme.typography.bodyMedium,
) {
    val parts = name.split(" ").map { word ->
        word.lowercase().replaceFirstChar { it.uppercaseChar() }
    }
    val text = buildAnnotatedString {
        parts.forEachIndexed { index, word ->
            if (index == parts.lastIndex && parts.size > 1) {
                withStyle(SpanStyle(fontWeight = FontWeight.SemiBold)) { append(word) }
            } else {
                append(word)
                if (index < parts.lastIndex) append(" ")
            }
        }
    }
    Text(
        text = text,
        modifier = modifier,
        style = style,
        maxLines = 1,
        overflow = androidx.compose.ui.text.style.TextOverflow.Ellipsis,
    )
}

private fun isPlaceholderUrl(url: String) =
    url.contains("cloud32.it/Associazioni/img/Uomo-1.png")
